src/ir/from_ast/astconv.rs

```diff
@@ -854,58 +883,72 @@ impl<'ctx, 'prog> BuildCtx<'ctx, 'prog> {
         cmds.extend(builder.port_assumptions());
         cmds.extend(body_cmds);
         ir_comp.cmds = cmds;
-        ir_comp
     }
 }
 
 pub fn transform(ns: ast::Namespace) -> ir::Context {
     let mut sig_map = SigMap::default();
+    let main_idx = ns.main_idx();
     let (mut ns, order) = crate::utils::Traversal::from(ns).take();
-    // Extract components in order
-    let components = order
-        .into_iter()
-        .map(|cidx| (cidx, std::mem::take(&mut ns.components[cidx])))
-        .collect_vec();
     // chains external signatures and component signatures (in post-order) to get all signatures associated with this namespace
-    let signatures = ns
-        .externals()
-        .map(|(_, sig)| sig)
-        .chain(components.iter().map(|(_, c)| &c.sig));
+    // let signatures = ns
+    //     .externals()
+    //     .map(|(_, sig)| sig)
+    //     .chain(components.iter().map(|(_, c)| &c.sig));
     // Walk over signatures and build a SigMap
-    for (idx, sig) in signatures.enumerate() {
+    for (idx, sig) in ns.signatures().map(|(_, sig)| sig).enumerate() {
         sig_map.insert(sig.name.copy(), Sig::from((sig, idx)));
     }
 
-    let main_idx = ns.main_idx();
     let mut ctx = ir::Context::default();
-    for (_, exts) in ns.externs {
+    for (file, exts) in ns.externs {
         for ext in exts {
             let idx = sig_map.get(&ext.name).unwrap().idx;
-            log::debug!("Compiling external {}: {}", ext.name, idx);
+            log::debug!("Converting external {}: {}", ext.name, idx);
             let ir_ext = BuildCtx::external(&ctx, ext);
             ctx.comps.checked_add(idx, ir_ext);
+            ctx.externals.entry(file.clone()).or_default().push(idx);
         }
     }
 
-    for (cidx, comp) in components {
-        log::debug!("Compiling component {}", comp.sig.name);
+    // declare all components
+    for (cidx, comp) in ns.components.iter().enumerate() {
         let idx = sig_map.get(&comp.sig.name).unwrap().idx;
-        let ir_comp = BuildCtx::comp(&ctx, comp, &sig_map);
+
+        let mut ir_comp = ir::Component::new(false);
         if Some(cidx) == main_idx {
             ctx.entrypoint = Some(idx);
+            ir_comp.src_info =
+                Some(ir::InterfaceSrc::new(comp.sig.name.copy()));
         }
         ctx.comps.checked_add(idx, ir_comp);
+        if Some(cidx) == main_idx {
+            ctx.entrypoint = Some(idx);
+        }
+    }
+
+    // create a dummy component to be swapped into the context
+    let mut dummy = ir::Component::new(false);
+    for cidx in order {
+        let comp = std::mem::take(&mut ns.components[cidx]);
+        log::debug!("Compiling component {}", comp.sig.name);
+        let idx = sig_map.get(&comp.sig.name).unwrap().idx;
+
+        std::mem::swap(ctx.get_mut(idx), &mut dummy);
+        BuildCtx::comp(&ctx, &mut dummy, comp, &sig_map);
+        // swap the component back into place
+        std::mem::swap(ctx.get_mut(idx), &mut dummy);
```

Comment:
This is a particularly gnarly sequence of `std::mem` operations. Can we remove this by carefully passing a `&mut` reference to. the component?